### PR TITLE
Improve Azure.VM.PublicKey and Azure.VMSS.PublicKey

### LINKED ISSUE #1704: The way of checking that a VM or VMSS has Linux doesn't cover all cases
The way PSRule currently checks if a VM or VMSS has Linux installed doesn't cover all cases: `storageProfile.osDisk.osType` and `virtualMachineProfile.osProfile.linuxConfiguration` could be undefined in an ARM template

This check is used for the preconditions of Azure.VM.PublicKey and Azure.VMSS.PublicKey

### PR DESCRIPTION
**Pull request by VeraBE**
## PR Summary

Closes #1704 

Improve Azure.VM.PublicKey and Azure.VMSS.PublicKey by checking if the VM(SS)s have Linux in a more reliable way: storageProfile.osDisk.osType and virtualMachineProfile.OsProfile.linuxConfiguration could be undefined in an ARM template

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: src/PSRule.Rules.Azure/rules/Azure.Common.Rule.ps1 [hunk 1]
```diff
@@ -178,15 +178,47 @@ function global:SupportsHybridUse {
     }
 }
 
-function global:IsLinuxOS {
+function global:IsLinuxOS { # TODO doesn't need to be global
+    [CmdletBinding()]
```
Comment (VeraBE) [inline src/PSRule.Rules.Azure/rules/Azure.Common.Rule.ps1:182]: This function doesn't need to be global but I didn't manage to make it work by declaring it as local

File: src/PSRule.Rules.Azure/rules/LinuxOffers.Rule.ps1 [hunk 2]
```diff
@@ -0,0 +1,4618 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+$global:LinuxOffers = @(
```
Comment (VeraBE) [inline src/PSRule.Rules.Azure/rules/LinuxOffers.Rule.ps1:4]: This is temporal until we can come up with a good way to keep the information up to date. Let me know if I should add the script I used to generate it somewhere in particular

File: src/PSRule.Rules.Azure/rules/Azure.Common.Rule.ps1 (hunk 1 above)
Comment (BernieWhite) [reply src/PSRule.Rules.Azure/rules/Azure.Common.Rule.ps1:182]: The discovery and caching process of rules requires a global scoping. Which is a little different then how a PowerShell script by itself would operate. Functions that are in a local or script scope are not available when the rules execute.

See [Define helper functions](https://microsoft.github.io/PSRule/v2/scenarios/azure-resources/azure-resources/#define-helper-functions).

File: src/PSRule.Rules.Azure/rules/LinuxOffers.Rule.ps1 (hunk 2 above)
Comment (BernieWhite) [reply src/PSRule.Rules.Azure/rules/LinuxOffers.Rule.ps1:4]: @VeraBE Works for me. Let's talk about how we can get the list of offers from ARM. Maybe from a PowerShell call.

Ultimately if we can generate it as JSON in a CI process that would be awesome. 

This is what we have done previously for provider data exports. 

https://github.com/Azure/PSRule.Rules.Azure/blob/main/scripts/providers.psm1
https://github.com/Azure/PSRule.Rules.Azure/blob/main/.github/workflows/providers.yml

**Review by BernieWhite — APPROVED** [review]
@VeraBE Nice work! All good to merge.

**Pull request merged**